src/engine/sparqlExpressions/StringExpressions.cpp

```diff
@@ -198,27 +198,24 @@ using IriOrUriExpression =
 using StrlenExpression =
     StringExpressionImpl<1, LiftStringFunction<decltype(strlen)>>;
 
-// LCASE
-[[maybe_unused]] auto lowercaseImpl =
-    [](std::optional<std::string> input) -> IdOrLiteralOrIri {
+template <auto toLowerOrToUpper>
+auto upperOrLowerCaseImpl =
+    [](std::optional<ad_utility::triple_component::Literal> input)
+    -> IdOrLiteralOrIri {
   if (!input.has_value()) {
     return Id::makeUndefined();
   } else {
-    return toLiteral(ad_utility::utf8ToLower(input.value()));
+    auto newContent = std::invoke(
+        toLowerOrToUpper, asStringViewUnsafe(input.value().getContent()));
+    input.value().replaceContent(newContent);
+    return LiteralOrIri(std::move(input.value()));
```

Comment:
this has three times `input.value()`.
With an initial
`auto& literal = input.value()` in the `else` case, you can make the code a bit more readable.